### PR TITLE
add my "hide shares" userscript to this extension

### PR DESCRIPTION
**Pull request by wowperfect**
hey yall, first PR to this project so let me know if that's improvements you'd like of course. feel free to contact me on discord if needed. 

tested on firefox on macOS. seems to work :)

P.S.
didn't mean to change appNav.css, woops

### CODE REVIEW

**Comment by enchanted-sword** [comment]
neat! apologies for the lack of documentation on the utilities, i usually just add them as i go.

chutils already has a feature that filters the dashboard in a similar way, but i do like the "controls at the top of the feed" concept for quickly toggling a filter on or off.
image
there's probably a way the two could be merged while keeping the best parts of both.

**Comment by wowperfect** [comment]
hmmm I'm not sure if it's necessary to combine both. 

I mainly don't want to blow either file out of proportion for maintenance reasons. this just adds a checkbox+label and some classes to the DOM and it feels decently complicated already. thoughts?

if you'd like to take over maintaining this piece of code, feel free to merge it how you'd like though

**Comment by enchanted-sword** [comment]
it wouldn't be much of a complexity increase for the existing filtering system, since it's just adding another filter case.
image
overall, it's a pretty light feature, and the existing preference system should also handle adding a checkbox on the timeline (as long as you set preferences in localStorage, it'll update on change) so i don't think that would be an issue.

if you'd like to take a shot at that, feel free! but i'd be happy to credit you for it either way, since it was your concept.

**Comment by wowperfect** [comment]
the current feature makes a toggle at the top of the feed to that shares and replies are hidden, but if someone you follow shares or replies their own post that doesn't get hidden. I don't know if that functionality is achievable with this component. but I can check tomorrow

I would implement it this once and probably would not want to continue maintenance of it, whereas if it were its own component i wouldn't mind maintaining it going forward. 

should I look into it?

edit: how would I add a checkbox on the timeline with just config files?

**Comment by wowperfect** [comment]
the place where this happens is here. if there is more than one handle, and the handles don't match, then hide the post. in this component can I access handles in the headerbar?
```javascript
function markShare(thread) {
  const $thread = $(thread)
  const handles = $.map($thread.find(handleSelector), x => x.href)
  if (handles.length > 1 && handles[0] !== handles[1]) {
    $thread.parent().addClass(isShareClass)
  }
}
```

hidden:
image

not hidden:
image

**Comment by enchanted-sword** [comment]
> the place where this happens is here. if there is more than one handle, and the handles don't match, then hide the post. in this component can I access handles in the headerbar?

via the `getViewModel` method, the module has access to the JSON representation of every item in the thread and the associated posting projects, so it would be as simple as comparing the root post project to the share tree projects.

> edit: how would I add a checkbox on the timeline with just config files?

it would have to be added in the feature itself, but i've been meaning to add an `update` export to that feature anyways, which would mean the checkboxes could just toggle the preference and it would toggle that filter only without reprocessing anything else.

> I would implement it this once and probably would not want to continue maintenance of it, whereas if it were its own component i wouldn't mind maintaining it going forward.

fair, apologies if i came off as saying "this is good, now just change it entirely". my thought process was that we've already written a module that filters dashboard posts, so it would make sense to combine them, while also adopting the in-timeline toggles, which are more convenient on mobile and especially in the PWA.

**Comment by wowperfect** [comment]
okay after some thinking I think this is probably possible then... I also remembered why I did it the way I did, which is that I also have made infinite scroll for cohost and I don't think this would be immediately compatible with infinite scroll, but mine is.

we could talk on discord about that if you'd like, I opened another issue about implementing that feature. there's probably a smarter way to implement it that makes it more compatible with chutils

> it would have to be added in the feature itself, but i've been meaning to add an update export to that feature anyways, which would mean the checkboxes could just toggle the preference and it would toggle that filter only without reprocessing anything else.

could you provide a code example of this or say where else it's used? I'm unsure of this API

File: src/scripts/filtering.css [hunk 1]
```diff
@@ -1 +1,5 @@
-.renderIfVisible:has([data-filtering="hidden"]) { display: none; }
\ No newline at end of file
+.renderIfVisible:has([data-filtering="hidden"]) { display: none; }
+
+.-chutils-remove-shares--hide-shares .-chutils-remove-shares--is-share {
```
Comment (enchanted-sword) [inline src/scripts/filtering.css:3]: is there a reason the same hidden attribute wouldn't also work for hiding shares?

File: src/scripts/filtering.js [hunk 2]
```diff
@@ -25,6 +40,14 @@ const filterPosts = async posts => {
     filterTags && (tags = [...model.tags, ...model.shareTree.flatMap(share => share.tags)].map(str => str.trim().toLowerCase()).join(' '));
     filterCws && (cws = [...model.cws, ...model.shareTree.flatMap(share => share.cws)].map(str => str.trim().toLowerCase()).join(' '));
 
+    if (doHideShares) {
+      const poster = model.postingProject.handle
+      const sharedFrom = model?.shareTree[model?.shareTree.length - 1]?.postingProject?.handle
+      if (!!sharedFrom && poster !== sharedFrom) {
+        $(post).parent().parent().addClass(isShareClass)
```
Comment (enchanted-sword) [inline src/scripts/filtering.js]: in essence of the first comment, setting `customAttribute` to `hidden` would accomplish the same thing

File: src/scripts/filtering.json [hunk 3]
```diff
@@ -11,6 +11,12 @@
         "name": "hide your own posts",
         "tooltip": "hide your own posts on the dashboard"
       },
+      "hideShares": {
+        "value": true,
```
Comment (enchanted-sword) [inline src/scripts/filtering.json]: default value should always be `false`

File: src/scripts/filtering.json [hunk 4]
```diff
@@ -11,6 +11,12 @@
...
+        "type": "toggle",
+        "name": "hide shares toggle",
+        "tooltip": "adds a toggle on home feed to hide shares of accounts you don't follow. by @wowperfect"
```
Comment (enchanted-sword) [inline src/scripts/filtering.json]: if it specifically hides shares of accounts you don't follow, wouldn't it need to check the following state to do so?

**Review by enchanted-sword — CHANGES_REQUESTED** [review]

File: src/scripts/filtering.css (hunk 1 above)
Comment (wowperfect) [reply src/scripts/filtering.css:3]: when the checkbox is clicked, the .-chutils-remove-shares--hide-shares class is toggled, which allows shares to be toggled on and off without mutating the DOM again

File: src/scripts/filtering.js (hunk 2 above)
Comment (wowperfect) [reply src/scripts/filtering.js]: then when the user unchecks the "hide shares" box, we'd have to somehow find all the shares and mutate the DOM to show them again

File: src/scripts/filtering.json (hunk 4 above)
Comment (wowperfect) [reply src/scripts/filtering.json]: no, because if the share-er matches the share-ee then you follow the person and they have shared their own post

Comment (wowperfect) [reply src/scripts/filtering.json]: maybe better phrasing would be "hide shares except of someone you follow sharing their own post" except that sounds a bit wordy to me. 

thoughts?

Comment (enchanted-sword) [reply src/scripts/filtering.json]: just "hide shares" might be the best way to describe it. you will still get some opaque and transparent shares, but they're either contentful or transparent shares of contentful posts still all from the same blog. 

in that regard, i forgot to add it during the review, but it currently seems to only check the current share with the last index in the `shareTree`, which means that resharing your share of someone else's post would cause it to still show up with shares hidden. i would do something like:
```js
if (model.shareTree.some(post => post.postingProject.handle !== poster)) $(post).parent().parent().addClass(isShareClass);
```

File: src/scripts/filtering.css (hunk 1 above)
Comment (enchanted-sword) [reply src/scripts/filtering.css:3]: oops, my bad, disregard

File: src/scripts/filtering.json (hunk 4 above)
Comment (wowperfect) [reply src/scripts/filtering.json]: oo good idea thx

File: src/scripts/filtering.js [hunk 5]
```diff
@@ -37,17 +59,52 @@ const filterPosts = async posts => {
...
+    parent.append(spacer)
+    parent.append(label)
+    parent.attr('class', 'mb-2 flex flex-col items-center lg:flex-row')
+
+    $('#' + checkboxId).on("change", () => {
+      $('main').toggleClass(hideSharesClass)
+    })
+  }
+
+  mutationManager.start(parentSelector, () => {
```
Comment (enchanted-sword) [inline src/scripts/filtering.js]: non-anonymous function for mutation callback

Comment (wowperfect) [reply src/scripts/filtering.js]: does my new solution work?? we need to have a closure over the label variable in that function

Comment (enchanted-sword) [reply src/scripts/filtering.js]: seems to work!

File: src/scripts/!features.json [hunk 6]
```diff
@@ -1,5 +1,6 @@
 [
   "horizontal",
+  "hideShares",
```
Comment (enchanted-sword) [inline src/scripts/!features.json]: remove now that it's part of filtering

File: src/scripts/filtering.js [hunk 7]
```diff
@@ -36,18 +58,57 @@ const filterPosts = async posts => {
   }
 };
 
+let putLabelInParent
+
 export const main = async () => {
-  ({ ownPosts, duplicatePosts, filterText, filterTags, filterCws } = await getOptions('filtering'));
+  ({ ownPosts, hideShares, duplicatePosts, filterText, filterTags, filterCws } = await getOptions('filtering'));
   [filterText, filterTags, filterCws] = [filterText, filterTags, filterCws].map(list => list.trim() === '' ? false : list.split(',').map(str => str.trim().toLowerCase()));
   if (!ownPosts && duplicatePosts === 'disabled' && !filterText && !filterTags && !filterCws) return;
```
Comment (enchanted-sword) [inline src/scripts/filtering.js]: `!hideShares` should be added to this list of checks, otherwise it won't run the feature if only hide shares is enabled

**Review by enchanted-sword — CHANGES_REQUESTED** [review]

**Review by enchanted-sword — APPROVED** [review]
